### ISSUE TITLE
How to train multiple LoRAs at once?

### ISSUE THREAD

**Issue by meteorlin** [huggingface/peft#631]
Hi! I would like to train multiple LoRAs at once (for some reason). Although `requires_grad` is True for all LoRA weight matrices, only the first LoRA weight matrix will calculate the gradient, and the others will not calculate the gradient - and will not be updated. How can I train them in one forward process?

1. I initialize multiple LoRAs using the `add_adapter()` method
```python
bert_path = "prajjwal1/bert-tiny"
rank = 8
LoRA_amount = 6

model = CustomBert.from_pretrained(bert_path)
peft_config = LoraConfig(
    inference_mode=False, 
    r=rank, 
    lora_alpha=32, 
    lora_dropout=0.1
)
model = PeftModel(model, peft_config, adapter_name="0")
for LoRA_index in range(1, LoRA_amount):
    model.add_adapter(str(LoRA_index), peft_config)
```
2. This is the printed model architecture
```
testModel(
  (model): PeftModel(
    (base_model): LoraModel(
      (model): CustomBert(
        (bert): BertModel(
          (embeddings): BertEmbeddings(
            (word_embeddings): Embedding(30522, 128, padding_idx=0)
            (position_embeddings): Embedding(512, 128)
            (token_type_embeddings): Embedding(2, 128)
            (LayerNorm): LayerNorm((128,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (encoder): BertEncoder(
            (layer): ModuleList(
              (0): BertLayer(
                (attention): BertAttention(
                  (self): BertSelfAttention(
                    (query): Linear(
                      in_features=128, out_features=128, bias=True
                      (lora_dropout): ModuleDict(
                        (0): Dropout(p=0.1, inplace=False)
                        (1): Dropout(p=0.1, inplace=False)
                        (2): Dropout(p=0.1, inplace=False)
                        (3): Dropout(p=0.1, inplace=False)
                        (4): Dropout(p=0.1, inplace=False)
                        (5): Dropout(p=0.1, inplace=False)
                      )
                      (lora_A): ModuleDict(
                        (0): Linear(in_features=128, out_features=16, bias=False)
                        (1): Linear(in_features=128, out_features=16, bias=False)
                        (2): Linear(in_features=128, out_features=16, bias=False)
                        (3): Linear(in_features=128, out_features=16, bias=False)
                        (4): Linear(in_features=128, out_features=16, bias=False)
                        (5): Linear(in_features=128, out_features=16, bias=False)
                      )
                      (lora_B): ModuleDict(
                        (0): Linear(in_features=16, out_features=128, bias=False)
                        (1): Linear(in_features=16, out_features=128, bias=False)
                        (2): Linear(in_features=16, out_features=128, bias=False)
                        (3): Linear(in_features=16, out_features=128, bias=False)
                        (4): Linear(in_features=16, out_features=128, bias=False)
                        (5): Linear(in_features=16, out_features=128, bias=False)
                      )
                      (lora_embedding_A): ParameterDict()
                      (lora_embedding_B): ParameterDict()
                    )
                    (key): Linear(in_features=128, out_features=128, bias=True)
                    (value): Linear(
                      in_features=128, out_features=128, bias=True
                      (lora_dropout): ModuleDict(
                        (0): Dropout(p=0.1, inplace=False)
                        (1): Dropout(p=0.1, inplace=False)
                        (2): Dropout(p=0.1, inplace=False)
                        (3): Dropout(p=0.1, inplace=False)
                        (4): Dropout(p=0.1, inplace=False)
                        (5): Dropout(p=0.1, inplace=False)
                      )
                      (lora_A): ModuleDict(
                        (0): Linear(in_features=128, out_features=16, bias=False)
                        (1): Linear(in_features=128, out_features=16, bias=False)
                        (2): Linear(in_features=128, out_features=16, bias=False)
                        (3): Linear(in_features=128, out_features=16, bias=False)
                        (4): Linear(in_features=128, out_features=16, bias=False)
                        (5): Linear(in_features=128, out_features=16, bias=False)
                      )
                      (lora_B): ModuleDict(
                        (0): Linear(in_features=16, out_features=128, bias=False)
                        (1): Linear(in_features=16, out_features=128, bias=False)
                        (2): Linear(in_features=16, out_features=128, bias=False)
                        (3): Linear(in_features=16, out_features=128, bias=False)
                        (4): Linear(in_features=16, out_features=128, bias=False)
                        (5): Linear(in_features=16, out_features=128, bias=False)
                      )
                      (lora_embedding_A): ParameterDict()
                      (lora_embedding_B): ParameterDict()
                    )
                    (dropout): Dropout(p=0.1, inplace=False)
                  )
                  (output): BertSelfOutput(
                    (dense): Linear(in_features=128, out_features=128, bias=True)
                    (LayerNorm): LayerNorm((128,), eps=1e-12, elementwise_affine=True)
                    (dropout): Dropout(p=0.1, inplace=False)
                  )
                )
                (intermediate): BertIntermediate(
                  (dense): Linear(in_features=128, out_features=512, bias=True)
                  (intermediate_act_fn): GELUActivation()
                )
                (output): BertOutput(
                  (dense): Linear(in_features=512, out_features=128, bias=True)
                  (LayerNorm): LayerNorm((128,), eps=1e-12, elementwise_affine=True)
                  (dropout): Dropout(p=0.1, inplace=False)
                )
              )
              (1): BertLayer(
                (attention): BertAttention(
                  (self): BertSelfAttention(
                    (query): Linear(
                      in_features=128, out_features=128, bias=True
                      (lora_dropout): ModuleDict(
                        (0): Dropout(p=0.1, inplace=False)
                        (1): Dropout(p=0.1, inplace=False)
                        (2): Dropout(p=0.1, inplace=False)
                        (3): Dropout(p=0.1, inplace=False)
                        (4): Dropout(p=0.1, inplace=False)
                        (5): Dropout(p=0.1, inplace=False)
                      )
                      (lora_A): ModuleDict(
                        (0): Linear(in_features=128, out_features=16, bias=False)
                        (1): Linear(in_features=128, out_features=16, bias=False)
                        (2): Linear(in_features=128, out_features=16, bias=False)
                        (3): Linear(in_features=128, out_features=16, bias=False)
                        (4): Linear(in_features=128, out_features=16, bias=False)
                        (5): Linear(in_features=128, out_features=16, bias=False)
                      )
                      (lora_B): ModuleDict(
                        (0): Linear(in_features=16, out_features=128, bias=False)
                        (1): Linear(in_features=16, out_features=128, bias=False)
                        (2): Linear(in_features=16, out_features=128, bias=False)
                        (3): Linear(in_features=16, out_features=128, bias=False)
                        (4): Linear(in_features=16, out_features=128, bias=False)
                        (5): Linear(in_features=16, out_features=128, bias=False)
                      )
                      (lora_embedding_A): ParameterDict()
                      (lora_embedding_B): ParameterDict()
                    )
                    (key): Linear(in_features=128, out_features=128, bias=True)
                    (value): Linear(
                      in_features=128, out_features=128, bias=True
                      (lora_dropout): ModuleDict(
                        (0): Dropout(p=0.1, inplace=False)
                        (1): Dropout(p=0.1, inplace=False)
                        (2): Dropout(p=0.1, inplace=False)
                        (3): Dropout(p=0.1, inplace=False)
                        (4): Dropout(p=0.1, inplace=False)
                        (5): Dropout(p=0.1, inplace=False)
                      )
                      (lora_A): ModuleDict(
                        (0): Linear(in_features=128, out_features=16, bias=False)
                        (1): Linear(in_features=128, out_features=16, bias=False)
                        (2): Linear(in_features=128, out_features=16, bias=False)
                        (3): Linear(in_features=128, out_features=16, bias=False)
                        (4): Linear(in_features=128, out_features=16, bias=False)
                        (5): Linear(in_features=128, out_features=16, bias=False)
                      )
                      (lora_B): ModuleDict(
                        (0): Linear(in_features=16, out_features=128, bias=False)
                        (1): Linear(in_features=16, out_features=128, bias=False)
                        (2): Linear(in_features=16, out_features=128, bias=False)
                        (3): Linear(in_features=16, out_features=128, bias=False)
                        (4): Linear(in_features=16, out_features=128, bias=False)
                        (5): Linear(in_features=16, out_features=128, bias=False)
                      )
                      (lora_embedding_A): ParameterDict()
                      (lora_embedding_B): ParameterDict()
                    )
                    (dropout): Dropout(p=0.1, inplace=False)
                  )
                  (output): BertSelfOutput(
                    (dense): Linear(in_features=128, out_features=128, bias=True)
                    (LayerNorm): LayerNorm((128,), eps=1e-12, elementwise_affine=True)
                    (dropout): Dropout(p=0.1, inplace=False)
                  )
                )
                (intermediate): BertIntermediate(
                  (dense): Linear(in_features=128, out_features=512, bias=True)
                  (intermediate_act_fn): GELUActivation()
                )
                (output): BertOutput(
                  (dense): Linear(in_features=512, out_features=128, bias=True)
                  (LayerNorm): LayerNorm((128,), eps=1e-12, elementwise_affine=True)
                  (dropout): Dropout(p=0.1, inplace=False)
                )
              )
            )
          )
          (pooler): BertPooler(
            (dense): Linear(in_features=128, out_features=128, bias=True)
            (activation): Tanh()
          )
        )
        (dropout): Dropout(p=0.1, inplace=False)
      )
    )
  )
  (classifier): Linear(in_features=128, out_features=3, bias=True)
)

```

**Comment by younesbelkada** [comment]
Hi @meteorlin 
Thanks for the issue, I don't think this is currently supported but I am not sure
Can you try to manually set `requires_grad=True` for all the adapters?

**Comment by DefinitlyEvil** [comment]
Hello is this issue being addressed? This should not be closed.

**Comment by BenjaminBossan** [comment]
I think the initial issue was that even if `requires_grad=True` on all LoRA adapters, since only one of them is used for inference, the others are not part of the graph and thus don't get any gradients. The code suggested should work but I think it is important to consider the use case.

If I'm not missing something, If we just add a bunch of LoRA weights on top of `result`, without a nonlinearity, it is mathematically not different from a single LoRA weight. To simplify, if our forward call is:

```python
def forward(self, x):
    x = x + self.weight_A + self.weight_B
    return x
```

We might as well only have `self.weight_A`.

**Comment by DefinitlyEvil** [comment]
@BenjaminBossan Oh wow, this is enlightening! if `r` is set to be equal, there's no point to add another adapter mathematical speaking?

**Comment by BenjaminBossan** [comment]
Well, I'm 95% sure :) It's a similar reasoning as when you have a simple feedforward net (MLP) without nonlinearities, it is mathematically equivalent to a single layer net.

A minor complication is that we don't simply add a weight but two weights that are multiplied, and we have dropout, which would be applied to each lora weight independently, but those are probably negligible.

> if `r` is set to be equal, there's no point to add another adapter mathematical

Can't really wrap my head around what different `r` values would mean, probably someone has to sit down and do the actual math :D

But coming back to what I said earlier, it really depends on the use case: If I have adapters X and Y, then I first train X on dataset 0, next Y and dataset 1, and then both on dataset 2, it could be true that X learns 0 and 2, and Y learns 1 and 2. But it could also be that dataset 2's updates are only learned by X and not Y or vice versa. Maybe that's fine, maybe not, hard to say without more context.

**Comment by DefinitlyEvil** [comment]
@BenjaminBossan 
For `r` I meant the LoRA rank. 
`W@x + lora1(x, rank=8) + lora2(x, rank=8)`
I think there's a chance that `lora2` might override the efforts of `lora1`? (catastrophic forgetting).

**Comment by DefinitlyEvil** [comment]
> > For `r` I meant the LoRA rank.
> 
> Yes, I got that. What I meant is that mathematically, I can't wrap my head around how `r` would change the computation :)

@BenjaminBossan I remember someone suggested that `<>` shapedlinear layers act like KV mappers, same `r` means same capacity.
